packages/go/bomenc/utf32.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2024 Specter Ops, Inc.
+//
+// Licensed under the Apache License, Version 2.0
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package bomenc
+
+import (
+	"bytes"
+	"errors"
+)
+
+// utf32ToUTF8 converts UTF-32 encoded bytes to UTF-8.
+//
+// Advantages of using bitwise operations for this conversion:
+//
+//  1. Efficiency: Bitwise operations are highly optimized at the hardware level,
+//     often executing in a single CPU cycle. This makes them faster than
+//     arithmetic operations or function calls, especially important when
+//     processing large volumes of text.
+//
+//  2. Direct byte manipulation: Bitwise operations allow us to work directly
+//     with the binary representation of the data. This is crucial for correctly
+//     interpreting UTF-32 encoded characters, which span four bytes.
+//
+//  3. Precision: When dealing with character encodings, every bit matters.
+//     Bitwise operations ensure we're interpreting the data exactly as intended,
+//     without any unintended modifications that could occur with higher-level operations.
+//
+//  4. Endianness handling: UTF-32 can be either big-endian or little-endian.
+//     Bitwise operations provide a clean and efficient way to handle both
+//     encodings with minimal code duplication.
+//
+//  5. Memory efficiency: By manipulating bits directly, we avoid the need
+//     for intermediate data structures or type conversions, which can be
+//     beneficial for memory usage, especially when processing large files.
+//
+//  6. Portability: Bitwise operations behave consistently across different
+//     hardware architectures, ensuring our code works reliably on various systems.
+//
+//  7. Educational value: Understanding and using bitwise operations provides
+//     insights into low-level data representation, which is valuable knowledge
+//     for any programmer working with different character encodings or
+//     binary protocols.
+func utf32ToUTF8(data []byte, bigEndian bool) ([]byte, error) {
```

Comment:
There are a lot of bitwise operations and other arcane components in this function. Is there truly no upstream stdlib or library function that does the same so we don't need to maintain this?

See: https://pkg.go.dev/golang.org/x/text/encoding/unicode/utf32